src/reportengine/configparser.py

```diff
@@ -132,6 +133,78 @@ def f_(self, val, *args, **kwargs):
 
     return f_
 
+def parse_record_(self, val):
+    """Template function for parsing <key>_recorded_spec_ from lockfile"""
+    return val
+
+def record_from_defaults(f):
+    f"""Decorator for recording default values. Given a key, for example
+    `filter_defaults` there might exist several specifications or specs of
+    defaults, each spec will have its own label e.g foo and bar. First a
+    parse function must be defined for the key and decorated with
+    `record_from_defaults`
+
+        @record_from_defaults
+        def parse_filter_defaults(self, spec):
+            return spec
+
+    Next a rule must exist for obtaining the defaults, the most simple rule
+    would check for a provided spec in a dictionary. These rules must be named
+    `{_defaults_token}<key>` e.g
+
+        def {_defaults_token}filter_defaults(self, spec):
+            _defaults = dict(eggs="spam", breakfast="continental")
+            return _defaults[spec]
+
+    these functions should be expanded with some error handling and can also
+    allow for lazy initialization for defaults which are more expensive to
+    compute.
+
+    Now the spec from the runcard is used to select some defaults for the given
+    key. The lockfile keeps track of what defaults are used by updating the input
+    with a nested mapping like
+
+        <key>_recorded_spec_:
+            <spec>:
+                defaults_mapping
+
+    The concept of a lockfile is that it can be used in lieu of a runcard and
+    produce the exact same configuration as when it was created, even if
+    the defaults have since been altered. In order to gain access to this
+    functionality a production rule needs to be added to the class which takes
+    <key> as an argument and <key>_recorded_spec_=None as an optional argument
+
+        def produce_filter_defaults_rules(
+            self, filter_defaults, filter_defaults_recorded_spec_=None
+        ):
+            if filter_defaults_recorded_spec_:
+                # do something with recorded defaults
+            else:
+                # do something with current loaded defaults (and update runcard)
+
+    """
+    @functools.wraps(f)
+    def f_(self, *args, **kwargs):
+        res = f(self, *args, **kwargs)
+        key = trim_token(f.__name__)
+        lockkey = key + "_recorded_spec_"
+
+        try:
+            mapping = getattr(self, f"{_defaults_token}{key}")(res)
+        except AttributeError:
+            raise ConfigError(f"No defaults have been set for key: {key}")
```

Comment:
I don't think this should be a config error. This is a program error and should be a hard crash.

Reply:
of course, if somebody has decorated the function it needs defaults..